link.go

```diff
@@ -285,6 +287,14 @@ Loop:
 	for {
 		select {
 		case <-ctx.Done():
+			// we are now in an inconsistent state, so kill the connection.
```

Comment:
Seems worth making a little function for this since it's also down below and their behavior is linked.